app/uk/gov/hmrc/gform/sharedmodel/formtemplate/BooleanExpr.scala

```diff
@@ -31,17 +37,96 @@ final case object IsFalse extends BooleanExpr
-          if FormDataHelpers.get(data, FormComponentId(fieldId)).contains(value) =>
-        true
-      case Equals(UserCtx(_), Constant(value)) if retrievals.affinityGroupName == value => true
-      case Or(expr1, expr2)                                                             => isTrue(expr1, data, retrievals) | isTrue(expr2, data, retrievals)
-      case And(expr1, expr2)                                                            => isTrue(expr1, data, retrievals) & isTrue(expr2, data, retrievals)
-      case IsTrue                                                                       => true
-      case _                                                                            => false
+      case Equals(field1, field2) => equals(prepopData(field1, retrievals, data), prepopData(field2, retrievals, data))
+      case GreaterThan(field1, field2) =>
+        greaterThan(prepopData(field1, retrievals, data), prepopData(field2, retrievals, data))
+      case GreaterThanOrEquals(field1, field2) =>
+        greaterThanOrEquals(prepopData(field1, retrievals, data), prepopData(field2, retrievals, data))
+      case LessThan(field1, field2) =>
+        lessThan(prepopData(field1, retrievals, data), prepopData(field2, retrievals, data))
+      case LessThanOrEquals(field1, field2) =>
+        lessThanOrEquals(prepopData(field1, retrievals, data), prepopData(field2, retrievals, data))
+      case Or(expr1, expr2)  => isTrue(expr1, data, retrievals) | isTrue(expr2, data, retrievals)
+      case And(expr1, expr2) => isTrue(expr1, data, retrievals) & isTrue(expr2, data, retrievals)
+      case IsTrue            => true
+      case _                 => false
+    }
+  }
+
+  private def equals(left: String, right: String): Boolean =
+    (toMaybeBigDecimal(left), toMaybeBigDecimal(right), left, right) match {
+      case (Some(l), Some(r), _, _) => (l == r)
+      case (_, _, l, r)             => (l == r)
+    }
+
+  private def greaterThan(left: String, right: String): Boolean =
+    (toMaybeBigDecimal(left), toMaybeBigDecimal(right), left, right) match {
+      case (Some(l), Some(r), _, _) => (l > r)
+      case (_, _, l, r)             => (l > r)
+    }
+
+  private def greaterThanOrEquals(left: String, right: String): Boolean =
+    (toMaybeBigDecimal(left), toMaybeBigDecimal(right), left, right) match {
+      case (Some(l), Some(r), _, _) => (l >= r)
+      case (_, _, l, r)             => (l >= r)
+    }
+
+  private def lessThan(left: String, right: String): Boolean =
+    (toMaybeBigDecimal(left), toMaybeBigDecimal(right), left, right) match {
+      case (Some(l), Some(r), _, _) => (l < r)
+      case (_, _, l, r)             => (l < r)
+    }
+
+  private def lessThanOrEquals(left: String, right: String): Boolean =
+    (toMaybeBigDecimal(left), toMaybeBigDecimal(right), left, right) match {
+      case (Some(l), Some(r), _, _) => (l <= r)
+      case (_, _, l, r)             => (l <= r)
```

Comment:
`equals`,  `greaterThan`, `greaterThanOrEquals`, `lessThan` or `lessThanOrEquals` differs only in operation used, we should refactor this to use only single implementation with operation as extra parameter.